csharp/src/Ice/Reference.cs

```diff
@@ -1524,192 +1318,138 @@ private void ClearConnection(Connection connection)
                 {
-                    // Get the proxy's endpoint or query the locator to get endpoints
-                    if (Endpoints.Count > 0)
-                    {
-                        endpoints = Endpoints;
-                    }
-                    else if (LocatorInfo != null)
-                    {
-                        (endpoints, cached) =
-                            await LocatorInfo.ResolveIndirectReferenceAsync(this, cancel).ConfigureAwait(false);
-                    }
+                    return false;
                 }
 
-                if (endpoints.Count == 0)
+                // With ice1 when secure endpoint is required filter out all non-secure endpoints.
+                if (Protocol == Protocol.Ice1 && PreferNonSecure == NonSecure.Never && !endpoint.IsAlwaysSecure)
                 {
-                    throw new NoEndpointException(ToString());
+                    return false;
                 }
 
-                // Apply overrides and filter endpoints
-                IEnumerable<Endpoint> filteredEndpoints = endpoints.Where(endpoint =>
+                // Check if the endpoint is compatible with the proxy invocation mode, Twoway and Oneway invocation
+                // modes require a non datagram endpoint, Datagram invocation mode requires a datagram endpoint, the
+                // other invocation modes (BatchOneway and BatchDagram) are not supported.
+                return InvocationMode switch
                 {
-                    // Filter out opaque endpoints
-                    if (endpoint is OpaqueEndpoint || endpoint is UniversalEndpoint)
-                    {
-                        return false;
-                    }
-
-                    // Filter out based on InvocationMode and IsDatagram
-                    switch (InvocationMode)
-                    {
-                        case InvocationMode.Twoway:
-                        case InvocationMode.Oneway:
-                        case InvocationMode.BatchOneway:
-                            if (endpoint.IsDatagram)
-                            {
-                                return false;
-                            }
-                            break;
+                    InvocationMode.Twoway or InvocationMode.Oneway => !endpoint.IsDatagram,
+                    InvocationMode.Datagram => endpoint.IsDatagram,
+                    _ => false
+                };
+            });
 
-                        case InvocationMode.Datagram:
-                        case InvocationMode.BatchDatagram:
-                            if (!endpoint.IsDatagram)
-                            {
-                                return false;
-                            }
-                            break;
+            var orderedEndpoints = Communicator.OrderEndpointsByTransportFailures(filteredEndpoints).ToList();
+            if (orderedEndpoints.Count == 0)
```

Comment:
It would be clearer to check for `filteredEndpionts.Count == 0` instead as here we give the impression that the ordering might actually remove some endpoints.

Reply:
Is it cheaper to do it on the list than in the enumerable.